spec-terminology.md

```diff
@@ -1,30 +1,75 @@
-behavior:
+Our usage of the term is generally the same as the [standard
+usage](https://en.wikipedia.org/wiki/Undefined_behavior) from other
+languages.
 
-#### LLVM-undefined behavior (or LLVM-UB)
+<a name="LLVM-UB"></a>
 
-Cases of known LLVM-UB are a specific subset of undefined behavior in general.
-Rust code with LLVM-UB will cause `rustc` to generate LLVM IR exhibiting
-undefined behavior.
+### LLVM-undefined behavior (LLVM-UB)
 
-This is distinct from the general case of Rust undefined behavior, in which
-it is unknown whether `rustc` will generate well-behaved LLVM IR.
+As a special case of undefined behavior, we use the phrase **LLVM
+undefined behavior** to indicate things that are considered undefined
+behavior by LLVM itself (as opposed to by the Rust compiler). There is
+no theoretical difference between UB and LLVM-UB -- both can cause
+arbitrary things to happen in your code. However, as a practical
+measure, LLVM-UB is worth distinguishing because it is much more
+*likely to* in practice.
+
+<a name="unspecified"></a>
+
+## Unspecified behavior
+
+We use the term "unspecified behavior" to refer to behavior that may
+vary across Rust releases, depending on what options are given to the
+compiler, or even -- in extreme cases -- across executions of the Rust
+compiler. However, unlike undefined behavior, the resulting execution
+is not completely undefined, and it must typically fall within some
+range of possibilities. Often, we will not specify precisely *how*
+something is implemented, but rather the patterns that must work.
+
+An example of "unspecified behavior" is the [layout for structs with
+no declared `#[repr]` attribute][ucg-struct].  This layout can and
+does change across Rust releases -- but of course within a given
+compilation, a struct must have *some* layout. Moreover, we guarantee
+that programs can (for example) use `sizeof` to determine the size of
+that layout, or access fields using Rust syntax like `foo.bar`. This
+requires the layout to be communicated in some fashion but doesn't
+specify how that is done.
+
+[ucg-struct]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/structs-and-tuples.md
+
+Our usage of the term is generally the same as the [standard
+usage](https://en.wikipedia.org/wiki/Unspecified_behavior) from other
+languages.
+
+<a name="TBD"></a>
+
+## To Be Defined Behavior (TBD)
+
+We refer to some behavior as **to be defined** to indicate that --
+while it is currently unspecified -- we *intend* to define that
+behavior at some point as part of this project group (though plans can
+change). This helps to define the scope of the group, but it also
+indicates behavior that you would be able to rely upon in the future
+in stable code. Note that TBD behavior is **still unspecified** until
+a formal decision is made, though, so if you rely on it today, your
+code cannot be considered stable Rust (even if it compiles on the
+stable compiler).
```

Comment:
> You and I have a different opinion here. I agree that in general it would be nice if TBD behavior remains on nightly, but I think it is acceptable for it to hit stable as well, and this has already happened in all sorts of areas -- the most notable case is the interaction of raw pointers, but there are plenty of other examples (e.g., the details of symbol visibility and shraed libraries is something we've tinkered with a few times).

I think the issue is that by TBD you refer to what I refer to as "unspecified behavior", where we say that this behavior is defined to do something that's not UB, but we don't say what that something is.

Right now, almost everything is unspecified in these documents, but some things are "unspecified by design" - we have looked at them, and this is something that we have decided that we don't want to specify, like the Rust unwinding ABI - and some are "unspecified because we haven't had the time to look at it yet". 

I was hoping that TBD could be used for these second type of unspecified things, so that we would at least look at them before stabilization, and say whether we want them to be undefined, unspecified, or something else. 

Maybe from this point-of-view it is clearer why I think that shipping a TBD is not a great idea. It would mean that we have recognized that there is something important enough to mention here as something to consider, but that we haven't considered it enough to even say that we provide no guarantees about it.

Reply:
So I think the full range of things are:

* Undefined behavior -- always a bug. As we discussed some in privmsg, the "miri virtual machine" panics here.
* Unspecified behavior -- the machine does not panic, but we don't say what happens. As you point out, there are a number of *reasons* we might not have said what happens yet:
    * We never want to specify this because we expect to want to change it. Example: Rust struct layout (without a repr attribute).
    * We expect eventually to specify this, but we haven't figure out just how it should be specified *yet*. This is what I am calling TBD, and I think you are too.

An orthogonal discussion is whether "TBD" behavior should ever be possible on stable. It seems clear that this is *suboptimal*, in that it leaves room for "de facto" dependencies. 

I have tried to summarize this in a [hackmd doc](https://hackmd.io/7ZcKCaRiRO2_I_yK8yeauQ).

Reply:
Resolving this concern based on [this comment](https://rust-lang.zulipchat.com/#narrow/stream/210922-wg-ffi-unwind/topic/tbd.20vs.20ub.20vs.20unspecified.20behavior/near/178398565).